common/functions/src/scalars/expressions/cast_from_string.rs

```diff
@@ -95,7 +95,18 @@ pub fn string_to_timestamp(date_str: impl AsRef<[u8]>, tz: &Tz) -> Option<DateTi
             let t = format!("1970-01-01 00:00:00{}", &s[19..]);
             tz.datetime_from_str(&t, "%Y-%m-%d %H:%M:%S%.f").ok()
         } else {
-            tz.datetime_from_str(s, "%Y-%m-%d %H:%M:%S%.f").ok()
+            match tz.datetime_from_str(s, "%Y-%m-%d %H:%M:%S%.f") {
+                Ok(dt) => {
+                    // convert timestamp less than `1000-01-01 00:00:00` to `1000-01-01 00:00:00`
+                    if dt.year() < 1000 {
+                        tz.datetime_from_str("1000-01-01 00:00:00", "%Y-%m-%d %H:%M:%S")
```

Comment:
How about using API like `DateTime::<Tz>::from_utc(NaiveDate::from_ymd(1000, 1, 1).and_hms(0, 0, 0), tz);` so that we don't need to parse from str?